### PR TITLE
Mesh_3 - Document Sizing_field_with_aabb_tree

### PR DESCRIPTION
**Pull request by janetournois**
## Summary of Changes

Document `Sizing_field_with_aabb_tree` for `edge_size`

## Release Management

* Affected package(s): Mesh_3
* Small feature [Sizing_field_with_AABB_tree](https://cgal.geometryfactory.com/CGAL/Members/wiki/Features/Small_Features/Sizing_field_with_AABB_tree), pre-approved @lrineau 17 April 2023
* Link to compiled documentation (obligatory for small feature)
  - [added example](https://cgal.github.io/7307/v2/Mesh_3/index.html#Mesh_3EdgesSizingField)
  - [Sizing_field_with_aabb_tree](https://cgal.github.io/7307/v2/Mesh_3/structCGAL_1_1Sizing__field__with__aabb__tree.html)
  - [Mesh Domain Fields section](https://cgal.github.io/7307/v2/Mesh_3/group__PkgMesh3DomainFields.html)

* License and copyright ownership: unchanged

### CODE REVIEW

File: Mesh_3/include/CGAL/Mesh_constant_domain_field_3.h [hunk 1]
```diff
@@ -28,27 +28,77 @@
 
 namespace CGAL {
 
+/*!
+* @ingroup PkgMesh3DomainFields
+* The class `Mesh_constant_domain_field_3` is a model of concept `MeshDomainField_3`. It provides
+* a constant field accessible using queries on 3D-points.
+*
+* The class `Mesh_constant_domain_field_3` can also be customized through `set_size()` operations to become
+* a piecewise constant field, i.e.\ a sizing field with a constant size on each subpart
```
Comment (albert-github) [inline Mesh_3/include/CGAL/Mesh_constant_domain_field_3.h:37]: `i.e.\ a sizing` why the `\`?

File: Mesh_3/include/CGAL/Mesh_constant_domain_field_3.h [hunk 2]
```diff
@@ -28,27 +28,77 @@
...
+* of the domain.
+*
+* @tparam Gt is the geometric traits class. It must match the type `Triangulation::Geom_traits`,
+* where `Triangulation` is the nested type of the model of `MeshComplex_3InTriangulation_3` used
+* in the meshing process.
+*
+* @tparam Index is the type of index of the vertices of the triangulation.
```
Comment (albert-github) [inline Mesh_3/include/CGAL/Mesh_constant_domain_field_3.h:44]: Shouldn't `Index` be `Index_`?

File: Mesh_3/include/CGAL/Sizing_field_with_aabb_tree.h [hunk 3]
```diff
@@ -165,7 +219,17 @@ struct Sizing_field_with_aabb_tree
...
+#endif // DOXYGEN_RUNNING
+
+public:
+  /// \name Operations
+  /// @{
 
+  /*!
+  * returns the value of the sizing field at the point `p`,
+  * assumed to be included in the input complex feature with dimension `dimension`
+  * and mesh subcomplex index `index`.
```
Comment (albert-github) [inline Mesh_3/include/CGAL/Sizing_field_with_aabb_tree.h:231]: Shouldn't ``index `index` `` be ``index `id` ``

File: Mesh_3/include/CGAL/Mesh_constant_domain_field_3.h (hunk 1 above)
Comment (afabri) [reply Mesh_3/include/CGAL/Mesh_constant_domain_field_3.h:37]: Should be a comma.

File: Mesh_3/include/CGAL/Mesh_constant_domain_field_3.h (hunk 2 above)
Comment (afabri) [reply Mesh_3/include/CGAL/Mesh_constant_domain_field_3.h:44]: There is a` typedef Index_  Index`

File: Mesh_3/include/CGAL/Sizing_field_with_aabb_tree.h [hunk 4]
```diff
@@ -17,30 +17,59 @@
...
-          typename Facet_patch_id_map_ = CGAL::Default
+namespace CGAL {
+/*!
+* @ingroup PkgMesh3DomainFields
+*
+* The class `Sizing_field_with_aabb_tree` is a model of concept `MeshDomainField_3`.
+* It provides a sizing field to be used in the meshing process of a polyhedral surface with features.
+*
+* At each query point, the field value is the minimum of the distances to the
+* surface patches it does not belong to, and the distances to the curves it does belong to.
```
Comment (lrineau) [inline Mesh_3/include/CGAL/Sizing_field_with_aabb_tree.h:40]: > and the distances to the curves it does belong to

What does it mean? As it is, the distance between a query point and a curve it belongs is 0, by definition of "belong to".

File: Mesh_3/doc/Mesh_3/Mesh_3.txt [hunk 5]
```diff
@@ -1013,7 +1013,20 @@ Right: the mesh with added 1D-features along sharp edges and edges shared
 by at least two of the input polyhedral surfaces forming the complex.
 \cgalFigureCaptionEnd
 
+\subsubsection Mesh_3EdgesSizingField Sizing field for feature edges
```
Comment (afabri) [inline Mesh_3/doc/Mesh_3/Mesh_3.txt:1016]:
```suggestion
\subsubsection Mesh_3EdgesSizingField Sizing Field for Feature Edges
```

File: Mesh_3/doc/Mesh_3/Mesh_3.txt [hunk 6]
```diff
@@ -1013,7 +1013,20 @@ Right: the mesh with added 1D-features along sharp edges and edges shared
...
 
+The following example shows how to generate a mesh from a polyhedral complex
+or a polyhedral surface with polyline features.
+Polyline features are covered by a set of so-called "protecting balls" which
```
Comment (afabri) [inline Mesh_3/doc/Mesh_3/Mesh_3.txt:1020]: Here we speak for the first time about "Protecting Balls" in the User Manual.

File: Mesh_3/doc/Mesh_3/Mesh_3.txt [hunk 7]
```diff
@@ -1013,7 +1013,20 @@ Right: the mesh with added 1D-features along sharp edges and edges shared
...
+sizes are driven by the size component of `Mesh_edge_criteria_3`.
+The ideal size of the protecting balls can be computed using
+`CGAL::Sizing_field_with_aabb_tree` that helps start the feature protection
```
Comment (afabri) [inline Mesh_3/doc/Mesh_3/Mesh_3.txt:1023]: Here we qualify with the namespace,  three lines above we don't  and just having a look in the User Manual I notice that we do not do it in a coherent way. I think it should be dropped everywhere. Unless it is `CGAL::Mesh_3::....` @sloriot ??

File: Mesh_3/include/CGAL/Mesh_constant_domain_field_3.h [hunk 8]
```diff
@@ -28,27 +28,77 @@
...
+  Mesh_constant_domain_field_3(const FT& size) : d_(size) {}
+
+  /// @}
+
+  /// \name Operations
+  /// @{
 
-  /// Returns size
+  /*!
+  * Returns size
```
Comment (MaelRL) [inline Mesh_3/include/CGAL/Mesh_constant_domain_field_3.h:100]:
```suggestion
  * Returns the size of query points of dimension `dim` and index `index`.
```

File: Mesh_3/include/CGAL/Sizing_field_with_aabb_tree.h [hunk 9]
```diff
@@ -17,30 +17,61 @@
...
+/*!
+* @ingroup PkgMesh3DomainFields
+*
+* The class `Sizing_field_with_aabb_tree` is a model of concept `MeshDomainField_3`.
+* It provides a sizing field to be used in the meshing process of a polyhedral surface with features.
+*
+* At each query point `p`, the field value is the minimum of the distances to
+*   - the surface patches `p` does not belong to, and
+*   - for each curve `C` it does belong to, the intersection points of `C` with
+*     plane orthogonal to `C` and passing through `p`.
```
Comment (MaelRL) [inline Mesh_3/include/CGAL/Sizing_field_with_aabb_tree.h:42]:
```suggestion
*     the plane orthogonal to `C` and passing through `p`.
```

Comment (MaelRL) [reply Mesh_3/include/CGAL/Sizing_field_with_aabb_tree.h:42]: "Curves" are sequences of line segments (i.e., is the plane always clearly defined)?

File: Mesh_3/include/CGAL/Sizing_field_with_aabb_tree.h [hunk 10]
```diff
@@ -17,30 +17,61 @@
...
+*
+* This sizing field is designed to be used for mesh complex edges,
```
Comment (MaelRL) [inline Mesh_3/include/CGAL/Sizing_field_with_aabb_tree.h:44]: "for edges of the mesh complex"?

File: Mesh_3/include/CGAL/Sizing_field_with_aabb_tree.h [hunk 11]
```diff
@@ -54,32 +85,57 @@ struct Sizing_field_with_aabb_tree
...
+    >::type;
+
+public:
+  /// \name Creation
+  /// @{
+
+  /*!
+  * Constructor for the sizing field.
+  * @param d the maximal value returned by `operator()`, corresponding
+  * to an upper bound on feature edges length in the output mesh.
```
Comment (MaelRL) [inline Mesh_3/include/CGAL/Sizing_field_with_aabb_tree.h:116]:
```suggestion
  * to an upper bound on feature edge lengths in the output mesh.
```

File: Mesh_3/include/CGAL/Mesh_constant_domain_field_3.h [hunk 12]
```diff
@@ -57,11 +107,15 @@ class Mesh_constant_domain_field_3
     return d_;
   }
 
-  /// Sets size at any point of dimension `dim` and index `index`.
+  /*!
+  * Sets the size such that `operator()` will return size `size`
+  * at any query point of dimension `dim` and index `index`.
```
Comment (MaelRL) [inline Mesh_3/include/CGAL/Mesh_constant_domain_field_3.h:112]:
```suggestion
  * Sets the size such that `operator()` for any query point of dimension `dim` and index `index` to `size`.
```

File: Mesh_3/doc/Mesh_3/Mesh_3.txt (hunk 6 above)
Comment (janetournois) [reply Mesh_3/doc/Mesh_3/Mesh_3.txt:1020]: It's actually described in [this section](https://doc.cgal.org/latest/Mesh_3/index.html#Mesh_3Protectionof0and1dimensionalExposed)

**Comment by lrineau** [comment]
And `CHANGES.md` should be amended.

File: Mesh_3/include/CGAL/Sizing_field_with_aabb_tree.h [hunk 13]
```diff
@@ -17,34 +17,65 @@
...
+* to the protecting balls placement algorithm, since it ensures that no
+* protecting ball will intersect a surface patch to which the
+* corresponding vertex does not belong.
+*
+* @tparam GeomTraits is the geometric traits class. It must match the type `Triangulation::Geom_traits`,
+* where `Triangulation` is the nested type of the model of `MeshComplex_3InTriangulation_3` used
+* in the meshing process.
+* @tparam MeshDomain is the type of the domain. It must be a model of `MeshDomainWithFeatures_3`,
+* and derive from `CGAL::Polyhedral_mesh_domain_with_features_3`
+
```
[the remark this answers is not in the document]
Comment (janetournois) [reply Mesh_3/include/CGAL/Sizing_field_with_aabb_tree.h]: indeed it does not add any more information

**Comment by sloriot** [comment]
Successfully tested in CGAL-5.6-Ic-241

**Pull request merged**